File: src/web/v71/schemas/system.py
"""System schemas (09_API_SPEC §9)."""

from __future__ import annotations

from datetime import datetime
from typing import Any, Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class WebsocketStatus(BaseModel):
    connected: bool
    last_disconnect_at: datetime | None = None
    reconnect_count_today: int = 0


class KiwoomApiStatus(BaseModel):
    available: bool
    rate_limit_used_per_sec: float = 0.0
    rate_limit_max: float = 4.5


class TelegramBotStatus(BaseModel):
    active: bool
    circuit_breaker_state: Literal["CLOSED", "OPEN", "HALF_OPEN"] = "CLOSED"


class DatabaseStatus(BaseModel):
    connected: bool
    latency_ms: int = 0


class MarketStatus(BaseModel):
    is_open: bool
    session: Literal["PRE", "REGULAR", "POST"] | None = None
    next_open_at: datetime | None = None
    next_close_at: datetime | None = None


class AccountSnapshot(BaseModel):
    """Real Kiwoom account total (kt00018 총평가금액, 5분 TTL cache).

    박스 wizard 가 비중 (%) 입력 시 실제 잔고 대비 매수 규모를 표시하기
    위해 노출. ``None`` 인 경우는 buy_executor 비활성 또는 키움 fetch 실패
    -- frontend 는 fallback 메시지 처리.
    """

    total_capital: float | None = None


SystemStatusLit = Literal["RUNNING", "SAFE_MODE", "RECOVERING"]


class SystemStatusOut(BaseModel):
    """09_API_SPEC §9.1."""

    status: SystemStatusLit
    uptime_seconds: int

    websocket: WebsocketStatus
    kiwoom_api: KiwoomApiStatus
    telegram_bot: TelegramBotStatus
    database: DatabaseStatus

    feature_flags: dict[str, bool]

    market: MarketStatus
    current_time: datetime

    # 비중 결정 UX (BoxWizard) 에 사용하는 실제 계좌 잔고. 노출 안 됐다면
    # 키움 미연결 또는 fetch 실패. 없을 시 None.
    account: AccountSnapshot = AccountSnapshot()


class SystemHealthOut(BaseModel):
    """09_API_SPEC §9.2."""

    status: Literal["healthy", "degraded"]
    checks: dict[str, Literal["ok", "fail"]]
    details: dict[str, str] | None = None


class SafeModeRequest(BaseModel):
    reason: str = Field(min_length=1, max_length=500)


class SafeModeResponse(BaseModel):
    safe_mode: bool
    entered_at: datetime | None = None
    resumed_at: datetime | None = None


class SystemRestartOut(BaseModel):
    """09_API_SPEC §9.5."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    restart_at: datetime
    recovery_completed_at: datetime | None
    recovery_duration_seconds: int | None
    reason: str | None
    reason_detail: str | None
    reconciliation_summary: dict[str, Any] | None
    cancelled_orders_count: int


TaskStatusLit = Literal["PENDING", "RUNNING", "COMPLETED", "FAILED"]


class AsyncTaskOut(BaseModel):
    """09_API_SPEC §9.6."""

    task_id: UUID
    type: str
    status: TaskStatusLit
    progress: int = 0
    started_at: datetime | None = None
    completed_at: datetime | None = None
    result: dict[str, Any] | None = None
    error: str | None = None


class BoxEntryMissResponse(BaseModel):
    """09_API_SPEC §9.7."""

    task_id: UUID
    checked_stocks: int = 0
    found_misses: int = 0

File: frontend/src/api/system.ts
// system API client (09_API_SPEC §9).

import {
  apiGet,
  apiGetList,
  apiPost,
  type ApiListResponse,
} from '@/lib/api';

export type SystemStatusLit = 'RUNNING' | 'SAFE_MODE' | 'RECOVERING';

export interface SystemStatusOut {
  status: SystemStatusLit;
  uptime_seconds: number;
  websocket: {
    connected: boolean;
    last_disconnect_at: string | null;
    reconnect_count_today: number;
  };
  kiwoom_api: {
    available: boolean;
    rate_limit_used_per_sec: number;
    rate_limit_max: number;
  };
  telegram_bot: {
    active: boolean;
    circuit_breaker_state: 'CLOSED' | 'OPEN' | 'HALF_OPEN';
  };
  database: {
    connected: boolean;
    latency_ms: number;
  };
  feature_flags: Record<string, boolean>;
  market: {
    is_open: boolean;
    session: 'PRE' | 'REGULAR' | 'POST' | null;
    next_open_at: string | null;
    next_close_at: string | null;
  };
  current_time: string;
  // 박스 wizard 비중 표시용 실제 키움 잔고 (kt00018 5분 TTL cache).
  // ``total_capital`` 이 ``null`` 이면 buy_executor 비활성 또는 키움
  // fetch 실패 -- frontend 는 fallback (1억) + "(추정)" 표시.
  account?: {
    total_capital: number | null;
  };
}

export interface SystemHealthOut {
  status: 'healthy' | 'degraded';
  checks: Record<string, 'ok' | 'fail'>;
  details?: Record<string, string>;
}

export interface SafeModeRequest {
  reason: string;
}

export interface SafeModeResponse {
  safe_mode: boolean;
  entered_at: string | null;
  resumed_at: string | null;
}

export interface SystemRestartOut {
  id: string;
  restart_at: string;
  recovery_completed_at: string | null;
  recovery_duration_seconds: number | null;
  reason: string | null;
  reason_detail: string | null;
  reconciliation_summary: Record<string, unknown> | null;
  cancelled_orders_count: number;
}

export interface AsyncTaskOut {
  task_id: string;
  type: string;
  status: 'PENDING' | 'RUNNING' | 'COMPLETED' | 'FAILED';
  progress: number;
  started_at: string | null;
  completed_at: string | null;
  result: Record<string, unknown> | null;
  error: string | null;
}

export const systemApi = {
  status(): Promise<SystemStatusOut> {
    return apiGet<SystemStatusOut>('/api/v71/system/status');
  },
  health(): Promise<SystemHealthOut> {
    return apiGet<SystemHealthOut>('/api/v71/system/health');
  },
  enterSafeMode(reason: string): Promise<SafeModeResponse> {
    return apiPost<SafeModeResponse, SafeModeRequest>(
      '/api/v71/system/safe_mode',
      { reason },
    );
  },
  resume(): Promise<SafeModeResponse> {
    return apiPost<SafeModeResponse>('/api/v71/system/resume');
  },
  restarts(
    params: { limit?: number; cursor?: string; from_date?: string } = {},
  ): Promise<ApiListResponse<SystemRestartOut>> {
    return apiGetList<SystemRestartOut>('/api/v71/system/restarts', { params });
  },
  task(id: string): Promise<AsyncTaskOut> {
    return apiGet<AsyncTaskOut>(`/api/v71/system/tasks/${id}`);
  },
};
